### ISSUE TITLE
Nalu-Wind GPU demonstration 

### ISSUE THREAD

**Issue by sayerhs** [Exawind/nalu-wind#370]
I am creating this issue to track the issues and related progress on running a Nalu-Wind simulation GPUs. Thanks to the work performed in #202, #238, #301, and #336 I was able to run a simple case both on host and device. However, I am not sure that I have the linear solver settings correct. Here is the output I am seeing. 

**Linear solver output on host**
```
Time Step Count: 1 Current Time: 0.125
 dtN: 0.125 dtNm1: 0.1 gammas: 1 -1 0
 Max Courant: 0 Max Reynolds: 0 (realm_1)
   Realm Nonlinear Iteration: 1/1

realm_1::advance_time_step()
NLI    Name           Linear Iter      Linear Res     NLinear Res    Scaled NLR
---    ----           -----------      ----------     -----------    ----------
1/2    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            9       3.08169e-08      6.506e-05             1
        ContinuityEQS          8       2.78823e-13    4.67656e-10             1
```

**Linear solver output on GPU**
```
Time Step Count: 1 Current Time: 0.125
 dtN: 0.125 dtNm1: 0.1 gammas: 1 -1 0
 Max Courant: 0 Max Reynolds: 0 (realm_1)
   Realm Nonlinear Iteration: 1/1

realm_1::advance_time_step()
NLI    Name           Linear Iter      Linear Res     NLinear Res    Scaled NLR
---    ----           -----------      ----------     -----------    ----------
1/2    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            9       3.08169e-08      6.506e-05             1
        ContinuityEQS         50       1.41185e-12    4.67656e-10             1
```

While the momentum solves proceed the same on both host and device, the pressure solve looks quite different. I am using the standard [milestone.xml](https://github.com/Exawind/nalu-wind/blob/master/reg_tests/xml/milestone.xml) and the linear solver settings in this [input file](https://github.com/Exawind/nalu-wind/blob/master/reg_tests/test_files/periodic3dEdge/periodic3dEdge.i). I haven't yet tested `mt_sgs` so the momentum solve is running exclusively on host? 

Cc: @alanw0 @overfelt @jhux2 @srajama1 @jrood-nrel

**Comment by jhux2** [comment]
@sayerhs The continuity solve is using Chebyshev smoothing, which should be the same on CPU and GPU.  (It's kernel is a sparse matvec.)  Is this on a single GPU?

**Comment by sayerhs** [comment]
@jhux2 yes. This is a single MPI rank/ single GPU run. 

Since I posted the original message, I notice that if I keep submitting the run again and again there are times where the GPU run will converge in the same number of iterations as the CPU run, but most of the time it will hit the max 50 iterations.

**Comment by jhux2** [comment]
That is interesting -- the matvec is deterministic, so there must be something else that is not.  If you turn on Belos and MueLu verbosity, are the linear iterations the same between CPU and GPU?  What simulation are you using?  @srajama1 @lucbv

**Comment by lucbv** [comment]
@sayerhs what version of Trilinos did you use for this run?
As @jhux2 mentioned, it would be good to get the output from MueLu, it is possible that new aggregation algorithms are being used in your simulation and these are non-deterministic.
If you have configured MueLu with the flags for kokkos refactor, these new algorithms will be used by default even if your xml file is unchanged...

**Comment by sayerhs** [comment]
@jhux2 @lucbv thanks for your comments. I am using the following commit of Trilinos for the GPU and CPU builds. I'll turn up verbosity and post the data here. 

```
*** Base Git Repo: Trilinos
68783a8e03 [Fri Sep 13 11:49:39 2019 -0600] <trilinos-autotester@trilinos.org>
Merge Pull Request #5894 from mhoemmen/Trilinos/Tpetra-Directory-verbose-output-
```

**Comment by sayerhs** [comment]
I do have the Kokkos refactor flags turned on: https://github.com/sayerhs/exawind-builder/blob/c5de8b507070cead69c3070285fd1620aae2b5ff/codes/trilinos.bash#L99-L100

**Comment by lucbv** [comment]
@sayerhs one thing that can be done initially to avoid having some recent changes in MueLu affect your runs is to use the following in your xml file:
```
<Parameter name="use kokkos refactor" type="bool" value="false"/>
```
this will ask MueLu to run without the refactored kokkos path thus providing a deterministic behavior.

**Comment by sayerhs** [comment]
Thanks @lucbv. I'll try that too.

**Comment by sayerhs** [comment]
@jhux2 @lucbv @alanw0 so it seems that muelu was working fine. There was a field sync issue on Nalu-Wind side, and somehow my solutions are not the same with `mt_sgs`, I haven't had a chance to verify the latter yet, but it is very likely that I am doing something silly that I need to fix. 

I was able to successfully run the ABL neutral edge simulation for multiple timesteps on the GPU and I am getting comparable results for the host-only run and the GPU runs. I am sharing three log files:

[abl_neutral_GPU_NP1.txt](https://github.com/Exawind/nalu-wind/files/3628447/abl_neutral_GPU_NP1.txt) – The GPU run with single MPI rank and 1 GPU. 
[abl_neutral_CPU_NP1.txt](https://github.com/Exawind/nalu-wind/files/3628445/abl_neutral_CPU_NP1.txt) – A single processor run that I am using to compare the linear solver iterations/residuals etc. with the GPU run
[abl_neutral_CPU_NP16.txt](https://github.com/Exawind/nalu-wind/files/3628446/abl_neutral_CPU_NP16.txt) – What we probably want to be comparing as far as runtime performance is concerned... running in parallel at about 25k DOFs for the Pressure Poisson solve. 

The runtime for 5 timesteps from the different runs:

```
CPU_NP16:   6.9453
CPU_NP1 : 185.3176
GPU_NP1 :  30.1772
```

so room for improvement.

**Comment by lucbv** [comment]
@sayerhs the file `abl_neutral_CPU_NP16.txt` seems to be the serial run while `abl_neutral_CPU_NP1.txt` seems to be running with 16 processors.

**Comment by sayerhs** [comment]
@lucbv argh. Sorry. But the information in the log files should be correct. I'll update the links.

**Comment by alanw0** [comment]
Those times seem a little strange. The CPU16 run seems more than 16 times faster than the CPU1 run. (185.32 / 6.95 = 26.7) Perhaps there is significant noise in the timings?

**Comment by jhux2** [comment]
@sayerhs Thanks for the logs.  In the future, can you enable the Trilinos Teuchos timers via the Nalu-Wind input deck?  Having those will potentially help in drilling down on bottlenecks.

**Comment by sayerhs** [comment]
@alanw0 You bring up a good point... I ran this case multiple times and I see that NP=1 and 2 have higher times than 4 and above... see below:



**NP = 16**
```
STKPERF: Total Time: 10.6668
STKPERF: Total Time: 10.8217
STKPERF: Total Time: 10.6625
STKPERF: Total Time: 10.7915
STKPERF: Total Time: 10.6885
STKPERF: Total Time: 10.7252
STKPERF: Total Time: 11.0010
STKPERF: Total Time: 10.8769
STKPERF: Total Time: 10.9235
STKPERF: Total Time: 10.5465
```

**NP = 8**
```
STKPERF: Total Time: 21.5600
STKPERF: Total Time: 20.8466
STKPERF: Total Time: 20.9479
STKPERF: Total Time: 21.7834
STKPERF: Total Time: 20.7215
STKPERF: Total Time: 20.8340
STKPERF: Total Time: 21.7479
STKPERF: Total Time: 20.9717
STKPERF: Total Time: 21.6262
STKPERF: Total Time: 20.7725
```

**NP = 4**
```
STKPERF: Total Time: 46.6679
STKPERF: Total Time: 46.6673
STKPERF: Total Time: 46.5580
STKPERF: Total Time: 46.2464
STKPERF: Total Time: 46.6641
STKPERF: Total Time: 46.3608
STKPERF: Total Time: 46.3976
STKPERF: Total Time: 46.3059
STKPERF: Total Time: 46.3310
STKPERF: Total Time: 46.6916
```

**NP = 2**
```
STKPERF: Total Time: 118.4804
STKPERF: Total Time: 118.0598
STKPERF: Total Time: 118.7922
STKPERF: Total Time: 118.8748
STKPERF: Total Time: 118.4835
STKPERF: Total Time: 118.1095
STKPERF: Total Time: 118.8335
STKPERF: Total Time: 118.4640
STKPERF: Total Time: 118.2355
STKPERF: Total Time: 118.6848
```

**NP = 1**
```
STKPERF: Total Time: 457.2730
STKPERF: Total Time: 452.8246
STKPERF: Total Time: 455.0502
STKPERF: Total Time: 455.6152
STKPERF: Total Time: 453.0859
```

**Comment by sayerhs** [comment]
@jhux2 do you mean `summarize_muelu_timer` in the linear solver section? That won't be very useful, as it is called in https://github.com/Exawind/nalu-wind/blob/672277aa9aeb4e3253951d121c8de72a935994ac/src/LinearSolver.C#L126 which will be only called once for this problem.

**Comment by lucbv** [comment]
@alanw0 there is a possibility that at certain sizes the problem starts fitting in faster memory which can increase the speedup beyond the expected linear scaling.

**Comment by sayerhs** [comment]
@jhux2 Here are the log files with muelu verbosity set to `high` and `Teuchos::TimeMonitor::summarize` called at the end of Nalu-Wind `main` function. I am using the standard [milestone.xml](https://github.com/Exawind/nalu-wind/blob/master/reg_tests/xml/milestone.xml), let me know if there are any other settings I should be trying out.

[log.gpu_baseline_verbose.txt](https://github.com/Exawind/nalu-wind/files/3632385/log.gpu_baseline_verbose.txt)
[log.gpu_mt_sgs_verbose.txt](https://github.com/Exawind/nalu-wind/files/3632386/log.gpu_mt_sgs_verbose.txt)

**Comment by sayerhs** [comment]
@jhux2 @lucbv I attempted to run this with multiple MPI ranks and I get a segfault at this line: 

https://github.com/Exawind/nalu-wind/blob/3bb5be8a0325ca33d5265062257f535798a49425/src/LinearSolver.C#L138

```
[r104u33:14463] *** Process received signal ***
[r104u33:14463] Signal: Floating point exception (8)
[r104u33:14463] Signal code: Integer divide-by-zero (1)
[r104u33:14463] Failing at address: 0x3b3b10b
```

**Comment by alanw0** [comment]
I'm going to do a little profiling of the cpu runs. Those times seem to indicate that we have something n-squared in local problem-size...

**Comment by jhux2** [comment]
> Here are the log files with muelu verbosity set to `high` and `Teuchos::TimeMonitor::summarize` called at the end of Nalu-Wind `main` function. I am using the standard [milestone.xml](https://github.com/Exawind/nalu-wind/blob/master/reg_tests/xml/milestone.xml), let me know if there are any other settings I should be trying out.

Thanks, these have the timers I was interested in.  Here are the relevant ones:

log.gpu_baseline_verbose.txt
```
Belos: ICGS[2]: Ortho (Inner Product)                                           0.3782 (302)               
Belos: ICGS[2]: Ortho (Norm)                                                    0.0070 (191)               
Belos: ICGS[2]: Ortho (Update)                                                  0.0587 (302)               
Belos: ICGS[2]: Orthogonalization                                               0.4497 (191)               
Belos: Operation Op*x                                                           1.8712 (231)               
Belos: Operation Prec*x                                                         5.9957 (191)               
Belos: PseudoBlockGmresSolMgr total solve time                                  8.6906 (40)                
Ifpack2::Chebyshev::apply                                                       0.1262 (606)               
Ifpack2::Chebyshev::compute                                                     0.0223 (3)                 
Ifpack2::Relaxation::apply                                                      5.7121 (90)                
Ifpack2::Relaxation::compute                                                    0.0517 (30)                
Ifpack2::Relaxation::initialize                                                 0.0000 (3)
```

log.gpu_mt_sgs_verbose.txt
```
Belos: ICGS[2]: Ortho (Inner Product)                                           0.4180 (346)               
Belos: ICGS[2]: Ortho (Norm)                                                    0.0077 (213)               
Belos: ICGS[2]: Ortho (Update)                                                  0.0686 (346)               
Belos: ICGS[2]: Orthogonalization                                               0.5005 (213)               
Belos: Operation Op*x                                                           0.1390 (253)               
Belos: Operation Prec*x                                                         0.9051 (213)               
Belos: PseudoBlockGmresSolMgr total solve time                                  1.8594 (40)                
Ifpack2::Chebyshev::apply                                                       0.1257 (606)               
Ifpack2::Chebyshev::compute                                                     0.0225 (3)                 
Ifpack2::Relaxation::apply                                                      0.6329 (112)               
Ifpack2::Relaxation::compute                                                    0.3998 (30)                
Ifpack2::Relaxation::initialize                                                 0.2299 (3)     
```

Some observations:

- MTSGS is about 10x faster than SGS
- Chebyshev should be preferred if it can be used
- Inner products and orthogonalization are about half of the Belos total solve time

**Comment by srajama1** [comment]
@jhux2 : Yay (on the MTGS apply) ! The cluster GS we are working on should be even better. Expect that to come in Kokkos 3.0.

I believe MTGS setup is slower, if we can avoid calling that multiple times on fixed graphs we can improve the total times.

The s-step might help with the orthogonalization times.

**Comment by srajama1** [comment]
@jhux2 : What is the Chebyshev comment based on ? Is there a log file that uses Chebyshev ?

**Comment by jhux2** [comment]
@srajama1 I was looking at the Chebyshev timers in the summary above, where the number of applies is higher but the apply time is lower.  However, Chebyshev is used in the pressure solve -- if @sayerhs was running monolithic momentum, then this isn't really a fair comparison.

**Comment by srajama1** [comment]
@jhux2 Got it ! In general, I agree with the comment that if Chebyshev works then the cost of each apply will be the fastest. @sayerhs is this is monolithic ?

**Comment by sayerhs** [comment]
@srajama1 yes this is the monolithic solve that I had set-up a while ago. I haven't switched to segregated yet, still trying to debug the crash I see on 2 mpi ranks when calling muelu setup.

**Comment by jhux2** [comment]
@sayerhs wrote:
> I haven't switched to segregated yet, still trying to debug the crash I see on 2 mpi ranks when calling muelu setup.

Let us know if you need help with this.

**Comment by sayerhs** [comment]
@srajama1 Yes to all your questions: UVM is on, CUDA launch blocking is set, using CUDA-aware OpenMPI for this run. This is the only error output I get: https://github.com/Exawind/nalu-wind/issues/370#issuecomment-533254720 I haven't been able to tell if I am passing any host pointer that is being accessed on device.

**Comment by srajama1** [comment]
My first guess is same as yours, that some host pointers are getting passed around. @jhux2 : Can we set env variables for debug outputs ?

**Comment by jhux2** [comment]
> Can we set env variables for debug outputs ?

@sayerhs Do you have verbosity set to `high` in MueLu's xml deck?   How much of the MueLu set output do you see?

**Comment by sayerhs** [comment]
@jhux2 Here is the output from verbosity set to high in MueLu's xml deck.

[log.gpu_np2.txt](https://github.com/Exawind/nalu-wind/files/3649382/log.gpu_np2.txt)

**Comment by sayerhs** [comment]
@alanw0 After fixing the performance issues of NGP field updates on CPU, here are the revised timings for the ABL neutral run (10 timesteps, with 4 Picard iterations per timestep). @jrood-nrel I think at this point this is ready for profiling if you want to start on that. 

|Run | Time (s) |
| --- | --- |
| cpu_np1  |157.2946 | 
| cpu_np2  | 88.3923 |
| cpu_np4  | 46.7837 |
| cpu_np8  | 25.0686 |
| cpu_np16 | 14.6100 |
| gpu_np1  | 56.4280 |

**Comment by alanw0** [comment]
@sayerhs excellent, those timings look good. It will be fun to see the results of the profiling, and start improving further.

**Comment by srajama1** [comment]
@sayerhs : What are the CPUs here ?

**Comment by sayerhs** [comment]
@srajama1 They are Intel Skylake processors (https://www.nrel.gov/hpc/eagle-system-configuration.html)

**Comment by srajama1** [comment]
If it is skylake, then pretty good numbers :)

**Comment by sayerhs** [comment]
@srajama1 thanks for confirming! By the way, any pointers on the MueLu error on multiple mpi ranks  with 1 GPU per rank? I posted the log file with the verbose MueLu logging a few posts above. Let me know if u should try something else. All the changes necessary to run the case has been now merged with master. So let me know if you want to try running this yourself.

**Comment by alanw0** [comment]
@sayerhs just curious, did you build the code that uses 'ngp::Field::swap' on a cpu build? There may be a problem there, because ngp::stkFieldAdapter doesn't have the swap method. This is fixed by the stk pull-request I currently have in the trilinos system, hopefully it will get through tonight or in the morning.

**Comment by sayerhs** [comment]
Yes. I hacked the NgpField.hpp file to add that to the StkFieldAdapter.

**Comment by srajama1** [comment]
This looks like the spgemm kernel or the transpose kernel was the last call, @jhux2 can you please confirm ? I don't know if there is a way to say we have returned from those calls.

We have fixed several problems related to Volta on Kokkos Kernels side, some of them have been patched to Trilinos, @brian-kelley will remember what is patched and what is not.

**Comment by sayerhs** [comment]
@srajama1 since this happens fairly quickly, I can load the kernel_logger and get the exact names of the kernrls, if that would help. Is there anything I've to keep in mind when using the logger for MPI runs? That is, is there a way to prefix which rank is having trouble?

**Comment by srajama1** [comment]
I don't know any way to print the rank with the kernel logger. AFAIK, it has no notion of MPI.

**Comment by brian-kelley** [comment]
@sayerhs @srajama1 I put my permanent bugfix for spgemm into Trilinos directly. So current Trilinos should have a full performance, correct version.

**Comment by sayerhs** [comment]
@brian-kelley thanks for the heads up. I'll pull the latest trilinos and attempt a run. I'll report back with log files here.

**Comment by jhux2** [comment]
@sayerhs I would like to run the exact code versions that you have. Where you using branches?  I can't find the Trilinos SHA in develop or master, nor the Nalu-Wind SHA in master.

**Comment by sayerhs** [comment]
The Nalu runs were performed before the code was merged into master as part of testing before submitting #381 

```
commit 8c99a85e5612e4bd2770e31110bd6a2ead2b16ee
Author: Shreyas Ananthan <shreyas@umd.edu>
Date:   Thu Sep 26 11:12:54 2019 -0600

    Updates to get working version of ABL simulation on GPU (#381)

    * Replace edge mass-flux calculation with NGP version in LowMachEqSys

    * Workaround to handle enthalpy/temperature updates for NGP

    * Convert periodic delta solution update to NGP

    * Convert UDiag initialization to NGP

    * Add field update logic for NGP runs

    * Fix update of states for NGP

    * Remove debugging print messages

    * Fix error in ngp::Field::swap usage
```

```
commit 3fb53ebdb0a4624418f1a84ec6dbbfcf7cd57b9c
Merge: b6d6f4b f916d26
Author: trilinos-autotester <trilinos-autotester@trilinos.org>
Date:   Tue Sep 24 11:55:24 2019 -0600

    Merge Pull Request #5974 from mayrmt/Trilinos/region-mg-remove-unused-includes

    Automatically Merged using Trilinos Pull Request AutoTester
    PR Title: MueLu: remove unused header inclusions
    PR Author: mayrmt
```

**Comment by sayerhs** [comment]
With the latest commit ec5e76ed059 and Trilinos 12.17-g60a8999136, `ablNeutralEdge` regression test input file should run unmodified on GPU for a single MPI rank. Previous runs had to remove the `fix_pressure_at_node` and `time-varying pressure forcing` to get it to run on GPUs. 

Even with this new Trilinos version, I am still seeing 

```
[r105u33:16420] *** Process received signal ***
[r105u33:16420] Signal: Floating point exception (8)
[r105u33:16420] Signal code: Integer divide-by-zero (1)
[r105u33:16420] Failing at address: 0x3cebf1b
```

errors within MueLu when attempting to run with 2 MPI ranks. @jhux2 @srajama1 would either of you be able to run this on SNL system and see if it is something I am doing that is silly? 

@alanw0 @lucbv IS there anything special we need to do to get TpetraSegregated running on GPUs? That also segfaults even on a single MPI rank run.

**Comment by srajama1** [comment]
@jhux2 was going to run this and @lucbv is also looking the kokkos refactor portion of MueLu. @jhux2  @lucbv : Let me know if this is a kernel issue.

**Comment by jhux2** [comment]
> would either of you be able to run this on SNL system and see if it is something I am doing that is silly?

@sayerhs Yes, that's part of why I'm building on waterman.

**Comment by sayerhs** [comment]
@jhux2 Thanks! I figured as much. 

Would it be possible for me to get access to `waterman`? I would like to help figuring out the build/run issues @jhux2 is having, and accelerate this process.

**Comment by alanw0** [comment]
@sayerhs I'll talk to our OAA and/or Paul and get a request put in to try to get you access.

**Comment by lucbv** [comment]
@sayerhs I did a GPU run of MueLu on a non-related problem yesterday and it worked OK.
That being said it was a very different setup that I am using for my refactor work.
I still have one feature that I would like to refactor and will try to tackle it today so I can look more directly at the problem you are facing on Monday.
My plan is to use summit to test MueLu on GPUs though.

**Comment by sayerhs** [comment]
@lucbv I have replicated the results from my NREL Eagle run on Summit also. You can see them in `${PROJWORK}/${EXAWIND_PROJECT_ID}/shreyas/exawind/abl_neutral_gpu` 

@alanw0 I am seeing this error when I attempt to run on 2 MPI ranks on ORNL Summit. I assume this is somewhere in STK or Tpetra graph? 
```
[g34n09:53117] *** An error occurred in MPI_Neighbor_alltoallv
[g34n09:53117] *** reported by process [2154496201,0]
[g34n09:53117] *** on communicator MPI COMMUNICATOR 13 DUP FROM 3
[g34n09:53117] *** MPI_ERR_ARG: invalid argument of some other kind
[g34n09:53117] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[g34n09:53117] ***    and potentially your MPI job)
[g34n09:53118] *** An error occurred in MPI_Neighbor_alltoallv
[g34n09:53118] *** reported by process [2154496201,1]
[g34n09:53118] *** on communicator MPI COMMUNICATOR 13 DUP FROM 3
[g34n09:53118] *** MPI_ERR_ARG: invalid argument of some other kind
[g34n09:53118] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[g34n09:53118] ***    and potentially your MPI job)
```

**Comment by alanw0** [comment]
@sayerhs That's strange. I've seen something like that when an MPI rank had no data, but I doubt if that's the case with your 2-proc run... In any case, you can turn off the mpi-neighbor stuff with the macro in stk_util/parallel/CommNeighbors.hpp (STK_MPI_SUPPORTS_NEIGHBOR_COMM) Which mpi implementation is on summit?

**Comment by sayerhs** [comment]
It is `spectrum-mpi/10.3.0.1-20190611` but seems like OpenMPI under the hood

**Comment by sayerhs** [comment]
Also @alanw0 that won't turn off the usage in the Tpetra graph creation that is in TpetraLinearSystem correct?

**Comment by alanw0** [comment]
Yes it would. The CommNeighbors class that TpetraLinearSystem uses would still work, but it would use different mpi calls, not the MPI_Neighbor_\* calls. That could potentially make it perform worse, although depending on the mpi implementation, I'm not convinced the neighbor stuff is always better.

Also: currently you have to edit the CommNeighbors.hpp header to turn that macro off. I'll make an update so that you can turn it off with a -D option on our compile line, or with a cmake option during the trilinos config step.

**Comment by sayerhs** [comment]
@jhux2 @srajama1 I spent some time trying to run the ABL neutral case on GPUs with multiple MPI ranks on SNL waterman system. 

Run environment on waterman
```
Currently Loaded Modulefiles:
  1) binutils/2.30.0
  2) gcc/7.2.0
  3) cuda/9.2.88
  4) java/ibm/sdk/8.0.0
  5) numa/2.0.11
  6) openmpi/3.1.1/gcc/7.2.0/cuda/9.2.88
  7) git/2.10.1
  8) ninja/1.7.2
  9) cmake-3.12.3-gcc-7.2.0-sf4aa36
 10) netlib-lapack-3.8.0-gcc-7.2.0-qtd4yan
```

I launched it with the following command line

```
mpiexec -np 2 ~/exawind/source/nalu-wind/build/naluX --kokkos-ndevices=2 -i abl_neutral.yaml -o log.gpu_np2
```

and get the following error: 
```
==== backtrace ====
 0 0x00000000000468e4 ucs_debug_cleanup()  ???:0
 1 0x0000000000024670 ucp_tag_offload_unexp_eager()  ???:0
 2 0x000000000006ab64 uct_mm_ep_am_bcopy()  ???:0
 3 0x0000000000025a78 ucp_tag_offload_unexp_eager()  ???:0
 4 0x000000000002ffdc ucp_tag_send_nbr()  ???:0
 5 0x000000000000538c mca_pml_ucx_send()  ???:0
 6 0x00000000000c29f8 MPI_Send()  ???:0
 7 0x0000000018fa1ac4 Teuchos::(anonymous namespace)::sendImpl<unsigned long>()  ???:0
 8 0x0000000018f79158 Teuchos::send<int, unsigned long>()  ???:0
 9 0x0000000017bc3d18 Tpetra::Distributor::doPosts<Kokkos::View<unsigned long const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned long*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, void, void> >()  ???:0
10 0x0000000017bc53dc Tpetra::Distributor::doPostsAndWaits<Kokkos::View<unsigned long const*, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0u> >, Kokkos::View<unsigned long*, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, void, void> >()  ???:0
11 0x0000000017c76ca4 Tpetra::DistObject<char, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::doTransferNew()  ???:0
12 0x0000000017c70068 Tpetra::DistObject<char, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::doTransfer()  ???:0
13 0x0000000017c6f67c Tpetra::DistObject<char, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::doExport()  ???:0
14 0x0000000010a299ec sierra::nalu::TpetraLinearSystem::loadComplete()  ???:0
15 0x00000000107420e4 sierra::nalu::EquationSystem::assemble_and_solve()  ???:0
16 0x00000000108370b0 sierra::nalu::MomentumEquationSystem::assemble_and_solve()  ???:0
17 0x0000000010836138 sierra::nalu::LowMachEquationSystem::solve_and_update()  ???:0
18 0x0000000010754570 sierra::nalu::EquationSystems::solve_and_update()  ???:0
19 0x0000000010969f9c sierra::nalu::Realm::advance_time_step()  ???:0
20 0x0000000010a13d30 sierra::nalu::TimeIntegrator::integrate_realm()  ???:0
21 0x00000000109b3ac0 sierra::nalu::Simulation::run()  ???:0
22 0x000000001015b500 main()  ???:0
23 0x0000000000025100 generic_start_main.isra.0()  libc-start.c:0
24 0x00000000000252f4 __libc_start_main()  ???:0
```

So now I have three different errors:

- NREL Eagle: Crashes in MueLu 
- ORNL Summit: Crashes in MPI alltoall neighborv
- SNL Waterman: Crashes in Tpetra

**Comment by srajama1** [comment]
@sayerhs : This is strange. Even the Waterman error is in an MPI call, MPI_send in this case. @lucbv has reproduced the Summit error and trying to see debug further. We will get back to you. One could turn debug and verbosity options to see what is going on here, that is what I think @lucbv is doing.

**Comment by sayerhs** [comment]
@srajama1 @lucbv the summit error can be worked around by following Alan's suggestions.

**Comment by brian-kelley** [comment]
I'm seeing the MPI_Neighbor_alltoallv errors in the STK unit tests. This is a local build with OpenMPI 4.0.1 and GCC 9.1, so I don't think the issue is specific to Summit or Spectrum.

```
*** Starting test BulkData.testFieldComm from UnitTestBulkData.cpp:2178
[s1042556:82567] *** An error occurred in MPI_Neighbor_alltoallv
[s1042556:82567] *** reported by process [1385037825,3]
[s1042556:82567] *** on communicator MPI COMMUNICATOR 4 DUP FROM 3
[s1042556:82567] *** MPI_ERR_ARG: invalid argument of some other kind
[s1042556:82567] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[s1042556:82567] ***    and potentially your MPI job)
[s1042556.srn.sandia.gov:82559] 1 more process has sent help message help-mpi-errors.txt / mpi_errors_are_fatal
[s1042556.srn.sandia.gov:82559] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

**Comment by lucbv** [comment]
@sayerhs as @srajama1 mentioned, I have a stack built on Summit and I have started with pure MueLu tests using a driver we have in Trilinos. This worked well enough, I was able to run that driver on up to 4 nodes.
I have now moved on to running the ablNeutralEdge so we can make some progress on the ExaWind front. As you mentioned earlier and as @brian-kelley is reporting, the communicator is generating the following error:
```
IOSS: Using decomposition method 'RCB' for 390,625 elements on 2 processors.
[h36n01:157558] *** An error occurred in MPI_Neighbor_alltoallv
[h36n01:157558] *** reported by process [2154496201,0]
[h36n01:157558] *** on communicator MPI COMMUNICATOR 13 DUP FROM 3
[h36n01:157558] *** MPI_ERR_ARG: invalid argument of some other kind
[h36n01:157558] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[h36n01:157558] ***    and potentially your MPI job)
[h36n01:157559] *** An error occurred in MPI_Neighbor_alltoallv
[h36n01:157559] *** reported by process [2154496201,1]
[h36n01:157559] *** on communicator MPI COMMUNICATOR 13 DUP FROM 3
[h36n01:157559] *** MPI_ERR_ARG: invalid argument of some other kind
[h36n01:157559] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[h36n01:157559] ***    and potentially your MPI job)
```

I will attempt modifying `stk_util/parallel/CommNeighbors.hpp` to go around this issue while we wait for the STK team to give us a better solution. Hopefully this will allow us to make progress on the linear solver/preconditioner. I spent a significant amount of time on this last week so I will have less time this week although I'll do my best to get the fix in place so I can get the ground running quickly next week.

**Comment by alanw0** [comment]
@lucbv and @brian-kelley I'll get an update in asap to turn off the mpi-neighbor stuff by default in stk until we can figure out what's going on there.

**Comment by lucbv** [comment]
@alanw0 I just added a couple of lines to undefine the variable for Neighbors_Comm and things are now running fine.
I'll keep an eye out for your fix and I'll update my Trilinos then. In the meantime it is not blocking the runs.

**Comment by alanw0** [comment]
@lucbv ok great.

**Comment by lucbv** [comment]
@sayerhs after doing the small fix in STK I was able to run the ablNeutralEdge on 2 GPU ang get timers out off MueLu.
Do you have specific settings that you want to use on Summit for ablNeutralEdge?

**Comment by sayerhs** [comment]
@lucbv The input file in the regression test is the right one. I just changed `sgs` to `mt_sgs` in the momentum/scalar solvers.

**Comment by lucbv** [comment]
@sayerhs OK, do we want to run as it is prescribed in the unit-test i.e. using 8mpi ranks with 1GPU per rank? That will push things into low occupancy for the GPU but it can be done for the sake of testing...

**Comment by sayerhs** [comment]
@lucbv I was planning to do 2 MPI ranks just to make sure everything is working. Beyond that we should use bigger meshes that is in solver performance repo.

**Comment by lucbv** [comment]
@sayerhs my run using 2GPUs on the same node did not crash anywhere, but I did get an odd behavior as the number of linear iterations needed to solve the continuity equations climbed up to 75 (the max set in the input deck). I am guessing that something goes wrong in the update of the solution after the linear problem is solved?
I will dig a bit more into it before delivering a final conclusion on that issue.

**Comment by sayerhs** [comment]
@lucbv ok. Can you share the Nalu residual output for the first timestep?

**Comment by lucbv** [comment]
@sayerhs sure, here it is, notice how the non-linear residual for enthalpy and continuity is increasing over the successive iterations:
```
*******************************************************
Simulation Shall Commence: number of processors = 2
*******************************************************
Realm::populate_variables_form_input() candidate input time: 0 for Realm: fluidRealm
Realm shall provide output files at : currentTime/timeStepCount: 0/0 (fluidRealm)
*******************************************************
Time Step Count: 1 Current Time: 0.5
 dtN: 0.5 dtNm1: 0.5 gammas: 1 -1 0
 Max Courant: 0.100746 Max Reynolds: 3.22386e+18 (fluidRealm)
   Realm Nonlinear Iteration: 1/1

fluidRealm::advance_time_step() 
NLI    Name           Linear Iter      Linear Res     NLinear Res    Scaled NLR
---    ----           -----------      ----------     -----------    ----------
1/4    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            4           24.1787         493647             1

*******************************************************
***** Belos Iterative Solver:  Pseudo Block Gmres 
***** Maximum Iterations: 75
***** Block Size: 1
***** Residual Test: 
*****   Test 1 : Belos::StatusTestImpResNorm<>: (2-Norm Res Vec) / (2-Norm Prec Res0), tol = 1e-06
*******************************************************
Iter  0, [ 1] :    1.000000e+00
Iter  1, [ 1] :    7.139635e-01
Iter  2, [ 1] :    2.597908e-01
Iter  3, [ 1] :    4.511084e-02
Iter  4, [ 1] :    8.471111e-03
Iter  5, [ 1] :    1.950707e-03
Iter  6, [ 1] :    9.453567e-04
Iter  7, [ 1] :    8.087680e-04
Iter  8, [ 1] :    5.624477e-04
Iter  9, [ 1] :    1.631327e-04
Iter 10, [ 1] :    3.387347e-05
Iter 11, [ 1] :    6.975861e-06
Iter 12, [ 1] :    1.475318e-06
Iter 13, [ 1] :    3.161466e-07
        ContinuityEQS         13         0.0238918        117.626             1
 1/1         myEnth
        EnthalpyEQS            3       2.85073e-06        77.0838             1
 1/1          myTke
        TurbKineticEnergyEQS   4       2.44166e-14      0.0183513             1
2/4    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            4         0.0833407        401.647   0.000813634

*******************************************************
***** Belos Iterative Solver:  Pseudo Block Gmres 
***** Maximum Iterations: 75
***** Block Size: 1
***** Residual Test: 
*****   Test 1 : Belos::StatusTestImpResNorm<>: (2-Norm Res Vec) / (2-Norm Prec Res0), tol = 1e-06
*******************************************************
Iter  0, [ 1] :    1.000000e+00
Iter  1, [ 1] :    2.367383e-01
Iter  2, [ 1] :    4.220717e-02
Iter  3, [ 1] :    7.518747e-03
Iter  4, [ 1] :    1.339148e-03
Iter  5, [ 1] :    2.489934e-04
Iter  6, [ 1] :    7.584913e-05
Iter  7, [ 1] :    5.624755e-05
Iter  8, [ 1] :    4.422120e-05
Iter  9, [ 1] :    1.670848e-05
Iter 10, [ 1] :    3.856710e-06
Iter 11, [ 1] :    8.238928e-07
        ContinuityEQS         11        0.00311189        5.87891     0.0499796
 1/1         myEnth
        EnthalpyEQS            3        0.00260494        595.299       7.72275
 1/1          myTke
        TurbKineticEnergyEQS   3        0.00096976        3.02644       164.917
3/4    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            5       5.56807e-05        12.8135   2.59568e-05

*******************************************************
***** Belos Iterative Solver:  Pseudo Block Gmres 
***** Maximum Iterations: 75
***** Block Size: 1
***** Residual Test: 
*****   Test 1 : Belos::StatusTestImpResNorm<>: (2-Norm Res Vec) / (2-Norm Prec Res0), tol = 1e-06
*******************************************************
Iter  0, [ 1] :    1.000000e+00
Iter  1, [ 1] :    2.330643e-01
Iter  2, [ 1] :    4.181194e-02
Iter  3, [ 1] :    7.424380e-03
Iter  4, [ 1] :    1.316250e-03
Iter  5, [ 1] :    2.340567e-04
Iter  6, [ 1] :    4.289792e-05
Iter  7, [ 1] :    1.264655e-05
Iter  8, [ 1] :    9.437028e-06
Iter  9, [ 1] :    7.490720e-06
Iter 10, [ 1] :    2.705274e-06
Iter 11, [ 1] :    6.055925e-07
        ContinuityEQS         11        0.00227134        5.83773     0.0496295
 1/1         myEnth
        EnthalpyEQS            3         0.0017004        333.606       4.32784
 1/1          myTke
        TurbKineticEnergyEQS   3       2.99145e-05       0.246526       13.4337
4/4    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            5       2.43023e-05        5.50835   1.11585e-05

*******************************************************
***** Belos Iterative Solver:  Pseudo Block Gmres 
***** Maximum Iterations: 75
***** Block Size: 1
***** Residual Test: 
*****   Test 1 : Belos::StatusTestImpResNorm<>: (2-Norm Res Vec) / (2-Norm Prec Res0), tol = 1e-06
*******************************************************
Iter  0, [ 1] :    1.000000e+00
Iter  1, [ 1] :    2.339403e-01
Iter  2, [ 1] :    4.197158e-02
Iter  3, [ 1] :    7.449281e-03
Iter  4, [ 1] :    1.320818e-03
Iter  5, [ 1] :    2.350754e-04
Iter  6, [ 1] :    4.290338e-05
Iter  7, [ 1] :    1.199509e-05
Iter  8, [ 1] :    8.657056e-06
Iter  9, [ 1] :    7.011808e-06
Iter 10, [ 1] :    2.725373e-06
Iter 11, [ 1] :    6.226763e-07
        ContinuityEQS         11        0.00262676        6.56599     0.0558208
 1/1         myEnth
        EnthalpyEQS            3        0.00187625         244.92       3.17733
 1/1          myTke
        TurbKineticEnergyEQS   3       3.45448e-06      0.0253713       1.38253
Mass Balance Review:  
Density accumulation: 0
Integrated inflow:    0
Integrated open:      0
Total mass closure:   0
Mean System Norm: 64.25503448183889 1 0.5
WallClockTime: 1 Pre: 0.0123949 NLI: 6.93144 Post: 0.0261681 Total: 6.97
```

**Comment by sayerhs** [comment]
@lucbv Thanks, I am still getting segfaults on Summit even after the STK fix, so it will take a while before I can get to this stage where I can look at residuals.

**Comment by sayerhs** [comment]
@lucbv I think your hunch is correct, we need to fix a few things on nalu-wind side to ensure parallel synchronization after linear solver update. I am attempting a rebuild of nalu-wind to see if I can fix the MPI errors before proceeding. Thanks for chasing this down.

**Comment by sayerhs** [comment]
@lucbv are you using a different branch of MueLu? My first Continuity Equation solve doesn't seem to have fully converged. See below. 

```
Time Step Count: 1 Current Time: 0.5
 dtN: 0.5 dtNm1: 0.5 gammas: 1 -1 0
 Max Courant: 0.100746 Max Reynolds: 3.22386e+18 (fluidRealm)
   Realm Nonlinear Iteration: 1/1

fluidRealm::advance_time_step()
NLI    Name           Linear Iter      Linear Res     NLinear Res    Scaled NLR
---    ----           -----------      ----------     -----------    ----------
1/1    Equation System Iteration
 1/1      myLowMach
        MomentumEQS            4            26.958         493647             1

*******************************************************
***** Belos Iterative Solver:  Pseudo Block Gmres
***** Maximum Iterations: 75
***** Block Size: 1
***** Residual Test:
*****   Test 1 : Belos::StatusTestImpResNorm<>: (2-Norm Res Vec) / (2-Norm Prec
Res0), tol = 1e-06
*******************************************************
Iter  0, [ 1] :    1.000000e+00
Iter  4, [ 1] :    6.540152e-02
Iter  8, [ 1] :    4.857854e-04
Iter 12, [ 1] :    1.809042e-05
Iter 14, [ 1] :    8.075963e-07
Belos::PseudoBlockGmresSolMgr::solve(): Warning! Solver has experienced a loss o
f accuracy!
        ContinuityEQS         14             91768        116.785             1
```

**Comment by sayerhs** [comment]
@lucbv #420 fixed an issue with field synchronization for shared nodes across MPI ranks after a linear solve. Can you please use the latest commit for your future tests?

**Comment by lucbv** [comment]
@sayerhs I will update my repo on Summit and see if I get better results.
I am also able to reproduce the issue in MueLu with the Transpose problem.
I should have an update on that soon!

**Comment by lucbv** [comment]
@sayerhs I am seeing an error in a CUDA debug build of Nalu-Wind on Summit. It seems to be in Nalu proper, I did rebuild with the latest version:

```
KokkosP: 18__nv_hdl_wrapper_tILb0ELb0E11__nv_dl_tagIPFvRKN3ngp10StaticMeshEN3stk8topology6rank_tERKNS5_4mesh8SelectorES_ILb0ELb0ES0_IMN6sierra4nalu33Turbulen\
ceAveragingPostProcessingEFvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKdSO_SO_S9_EXadL_ZNSE_33compute_temperature_resolved_fluxESM_SO_SO_SO_S9_E\
ELj1EEFvRKNS1_15StaticMeshIndexEEJKNS1_11StaticFieldIdEEKiSY_SY_SX_SX_SN_SN_SN_SN_EEEXadL_ZNSD_8nalu_ngp20run_entity_algorithmIS2_S10_EEvRKT_S7_SB_T0_EELj1EE\
FvRKN6Kokkos4Impl14CudaTeamMemberEEJKNS5_9NgpVectorIjEES3_KS7_KS10_EE
:0: : block: [589,0,0], thread: [0,0,0] Assertion `View bounds error of view deviceData_temperature` failed.
```

**Comment by sayerhs** [comment]
@lucbv Just pushed a bugfix, can you try with this? 

Curious why the address sanitizer didn't catch this. @jrood-nrel any thoughts?

**Comment by lucbv** [comment]
@sayerhs I will do a rebuild before going home, these debug builds are somewhat slow...

**Comment by lucbv** [comment]
@sayerhs the last run was clean and the convergence of the nonlinear and linear solvers looks good to me.
[ablNeutralEdge_1GPU.log](https://github.com/Exawind/nalu-wind/files/3779228/ablNeutralEdge_1GPU.log)

**Comment by jrood-nrel** [comment]
ASAN seemed to catch the last test error I put in the unit tests. Was this error able to be caught with valgrind? This function must act a little differently on the host? I wonder how ASAN does with all the Kokkos abstraction in general.

**Comment by sayerhs** [comment]
@lucbv I am getting correct results when running on 1GPU, the Belos solver error seems to happen when I run on 2MPI ranks with 1 GPU per rank.

**Comment by lucbv** [comment]
@sayerhs: noted, I was just checking that the patch you pushed on Friday did fix on Summit the `memory out of bound` I reported earlier with a debug build.
I can now push though the whole simulation using the debug build which is helpful!
I am still working on the 2 GPUs issue and making progress on isolating the exact source of the problem

**Comment by lucbv** [comment]
@sayerhs  a "quick fix" has been applied in my local source of Trilinos on Summit, the Trilinos team is working on a permanent fix that will be incorporated in github soon.
With this fix I am able to run the `ablNeutralEdge` to completion using 2 nodes with 4 GPUs per node (see [output](https://github.com/Exawind/nalu-wind/files/3810589/ablNeutralEdge_8GPU_temp_fix.log)).
However I am seeing a degraded linear solver performance after the first time-step, I am not sure why? Is it possible that something still needs to be updated in Nalu-Wind for GPU runs?
My Nalu-Wind SHA1 is: `761b1dd`, I'll try to update to something more recent soon

**Comment by lucbv** [comment]
@sayerhs  a "quick fix" has been applied in my local source of Trilinos on Summit, the Trilinos team is working on a permanent fix that will be incorporated in github soon.
With this fix I am able to run the `ablNeutralEdge` to completion using 2 nodes with 4 GPUs per node (see [output](https://github.com/Exawind/nalu-wind/files/3810589/ablNeutralEdge_8GPU_temp_fix.log)).
However I am seeing a degraded linear solver performance after the first time-step, I am not sure why? Is it possible that something still needs to be updated in Nalu-Wind for GPU runs?
My Nalu-Wind SHA1 is: `761b1dd`, I'll try to update to something more recent soon...

**Comment by sayerhs** [comment]
@lucbv thanks! If you can share the quick fix with me, I can look to see what's happening in the second timestep

**Comment by sayerhs** [comment]
@lucbv I reran this case, and I notice that the code works fine if I force reinitialization of the MueLu instance at the beginning of every timestep. See log file for the 4 GPU run [gpu_np4.txt](https://github.com/Exawind/nalu-wind/files/3815211/gpu_np4.txt)

**Comment by sayerhs** [comment]
@lucbv This is what I mean by forcing a reinit 

https://github.com/Exawind/nalu-wind/blob/d3bc42da8d293d28a2ebd963e7756b78c49e6f5c/src/LinearSolver.C#L136-L145

As long as we hit line 138 or 141, the code seems to converge properly. This isn't case with CPU only runs.

**Comment by lucbv** [comment]
@sayerhs OK, that's interesting.
Do we expect the `matrix_` object to be different between two time-steps?
I am guessing that since we are doing non-linear iteration that should be the case?

**Comment by sayerhs** [comment]
Yes, the diagonal term in the matrix will be different between the first and second timesteps, but somehow this doesn't seem to affect the CPU run.

**Comment by lucbv** [comment]
I see, there could be different default parameters used between CPU and GPU runs since the algorithms themselves are a bit different.
I will start by checking that and move from there. This does not look like a CS issue but more a Math or settings issue.

**Comment by lucbv** [comment]
@sayerhs just to make sure I am on the right track I took 5 minutes to run the problem with Gauss-Seidel instead of Chebyshev as my smoother and MueLu was converging in about 10 iterations at each time step.
I will try to tune the Chebyshev parameters to get back the CPU performance.

**Comment by sayerhs** [comment]
Successfully demonstrated with the regression test `ablNeutralEdge` on multiple MPI ranks with 1 GPU per rank.
